File: Catalogo_Grupo_4/CodigosPython/Jacobi.py
###############################################################################
import numpy as np
###############################################################################
def jacobi(A, b, x0, tol):
    '''
    Metodo de Jacobi
    :param A: matriz de coeficientes
    :param b: vector de terminos independientes
    :param x0: vector de valores iniciales
    :param tol: tolerancia de la respuesta
    :return: x: vector solucion
    '''
    n = len(A)
    A = np.array(A)
    b = np.array(b)
    L = np.zeros((n, n))
    D = np.zeros((n, n))
    U = np.zeros((n, n))

    for i in range(n):
        for j in range(n):
            if i > j:
                L[i][j] = A[i][j]
            elif i < j:
                U[i][j] = A[i][j]
            else:
                D[i][j] = A[i][j]

    x = np.array(x0)
    error = np.linalg.norm(np.dot(A, x) - b)
    i = 0
    m_inv = np.linalg.inv(D)
    m_times_n = np.dot(m_inv, (-1 * (L + U)))
    m_times_b = np.dot(m_inv, b)

    while error >= tol:
        x = np.dot(m_times_n, x) + m_times_b
        error = np.linalg.norm(np.dot(A, x) - b)
        i += 1

    return x

File: Catalogo_Grupo_4/CodigosC++/Jacobi.cpp
#include <iostream>
#include <armadillo>

using namespace std;
using namespace arma;

/**
 * @param A: matriz de coeficientes
 * @param b: vector de terminos independientes
 * @param xInicial: vector de valores iniciales
 * @param MAXIT: cantidad de iteraciones maximas
 * @param TOL: tolerancia de la respuesta
 * @return tuple<vec, double>: vector solucion, error de la solucion 
 */
tuple<vec, double> jacobi(mat A, vec b, vec xInicial, int MAXIT, double TOL)
{

    mat D(size(A), fill::zeros);
    mat U(size(A), fill::zeros);
    mat L(size(A), fill::zeros);

    for (int i = 0; i < A.n_rows; i++)
    {
        for (int j = 0; j < A.n_cols; j++)
        {
            if (j < i)
            {
                L(i, j) = A(i, j);
            }
            else if (j > i)
            {
                U(i, j) = A(i, j);
            }
            else if (i == j)
            {
                D(i, j) = A(i, j);
            }
            else
            {
                cout << "Error" << endl;
            }
        }
    }

    vec xk = xInicial;
    vec xk1;
    int iter = 0;
    double err = TOL + 1;

    while (iter < MAXIT)
    {
        xk1 = ((-D.i()) * (L + U) * (xk)) + ((D.i()) * (b));
        xk = xk1;
        err = norm(A * xk - b);

        if (err < TOL)
        {
            break;
        }
        else
        {
            iter = iter + 1;
        }
    }
    return make_tuple(xk, err);
}

/**
 * Ejemplo numerico
 */
int main()
{
    tuple<vec,double> testJ=jacobi("5 1 1;1 5 1;1 1 5","7 7 7","0 0 0",100,0.000001);
    cout << "Aproximacion: \n"
         << get<0>(testJ) << endl;
    cout << "Error: " << get<1>(testJ) << endl;
    return 0;
}